### ISSUE TITLE
Cannot add types-only dependencies that have exports (but none for package.json)

### ISSUE THREAD

**Issue by skleinjung** [projen/projen#2674]
Hello! I encountered difficulties installing a node package as a devDependency via projen today and was able to track the cause down to how tryResolveDependencyVersion works. The package I was unable to install is `@types/js-yaml`. The reason for my issue is as described below.

#### Expected Behavior

Adding `@types/js-yaml` as a dependency to a projen Node.js project would result in the package being installed.

#### Actual Behavior

- Error message during project synthesis: `unable to resolve version for @types/js-yaml from installed modules`
- The package is not present in node_modules or package.json

#### Details

The package.json for `@types/js-yaml` includes the following content:

```
    "exports": {
        ".": {
            "types": {
                "import": "./index.d.mts",
                "default": "./index.d.ts"
            }
        }
    }
```

The details of why this is failing:

- there is an "exports" field, but no export for package.json
- the package does not define any of the [standard export conditions](https://nodejs.org/api/packages.html#conditional-exports): `node-addons`, `node`, `import`, `require`, or `default`

Since this package only defines Typescript types, the only export is for `types`, which does not resolve by `require.resolve` unless `--conditions=types` is passed to the Node.js process. As far as I can tell, there is no way to override this at runtime.

The only way I was able to move forward was to use [patch-package](https://www.npmjs.com/package/patch-package) to apply the following patch to the `@types/js-yaml@4.0.5` package:

```
diff --git a/node_modules/@types/js-yaml/package.json b/node_modules/@types/js-yaml/package.json
index 1d8d7e9..1f29c78 100755
--- a/node_modules/@types/js-yaml/package.json
+++ b/node_modules/@types/js-yaml/package.json
@@ -48,6 +48,7 @@
                 "import": "./index.d.mts",
                 "default": "./index.d.ts"
             }
-        }
+        },
+        "./package.json": "./package.json"
     }
 }
\ No newline at end of file
```

This adds an export for `package.json`, which causes the current logic in projen to work. Unfortunately, the only way I could think of to resolve this is to have a directory-based fallback to resolve such packages (i.e. something gross like `const pathToPackageJson = path.resolve(process.cwd(), 'node_modules', 'some-package', 'package.json');` This will likely break with package managers that do fancy things with module resolution (like Yarn 2), and so is probably not a great general solution.

If there are ideas on other ways this could be approached, I could try and provide an implementation.

**Comment by mrgrain** [comment]
Thanks for reporting this issue!

We already had a similar issue with exports on regular packages that was fixed in [v0.71.51](https://github.com/projen/projen/releases/tag/v0.71.51).

Can your confirm if the issue here is still present in current versions?

**Comment by skleinjung** [comment]
Thanks for the quick reply! I am using v0.71.53. The fix you mentioned
works if the package has a default export. In this case ***@***.***/js-yaml) it
is a package that only contains Typescript types and has no "default"
export -- only a "types" export. This type of package has no runtime
component and so "require.resolve" will fail to resolve it.

If I can provide further information let me know.

**Comment by mrgrain** [comment]
@BradenM FYI - and if you happen to have the capacity I'd definitely love another contribution from you 😃

**Comment by BradenM** [comment]
@mrgrain Certainly :smile:  (see: #2681)

From what I was able to find, it seems @skleinjung is most likely correct that a directory/search fallback is the best/only option here.

One alternative I came up with but didn't pursue would be to spawn a node sub-process with the appropriate `types` condition that could execute the `require.resolve`. This would probably resolve the majority of scenarios here, but technically conditions can be named anything, so it would be far from perfect.

Seems that other libraries (https://github.com/wooorm/import-meta-resolve/blob/7ddaba4d809910db559107e39acf38782d86008d/lib/resolve.js#L967-L970) implement a similar fallback as well… so I ended up taking that approach.

**Comment by mrgrain** [comment]
@skleinjung Can you confirm if the latest release fixes this issue for you>?

**Comment by skleinjung** [comment]
This is fixed for me in v0.71.60.

I am using `npm` as package manager. As I mentioned in the initial report, this approach might not work for all package managers. In those cases it might be necessary to pin the package version in the projen project config.
